toolchain/check/check.cpp

```diff
@@ -656,6 +657,43 @@ auto NodeIdTraversal::Next() -> std::optional<Parse::NodeId> {
   }
 }
 
+// Emits a diagnostic for each declaration in context.definitions_required()
+// that doesn't have a definition.
+static auto DiagnoseMissingDefinitions(Context& context,
+                                       Context::DiagnosticEmitter& emitter)
+    -> void {
+  CARBON_DIAGNOSTIC(MissingDefinitionInImpl, Error,
+                    "No definition found for declaration in impl file");
+  for (SemIR::InstId decl_inst : context.definitions_required()) {
+    CARBON_KIND_SWITCH(context.insts().Get(decl_inst)) {
+      case CARBON_KIND(SemIR::ClassDecl class_decl): {
+        if (!context.classes().Get(class_decl.class_id).is_defined()) {
+          emitter.Emit(decl_inst, MissingDefinitionInImpl);
+        }
+        break;
+      }
+      case CARBON_KIND(SemIR::FunctionDecl function_decl): {
+        if (context.functions().Get(function_decl.function_id).definition_id ==
+            SemIR::InstId::Invalid) {
+          emitter.Emit(decl_inst, MissingDefinitionInImpl);
+        }
+        break;
+      }
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        if (!context.impls().Get(impl_decl.impl_id).is_defined()) {
+          emitter.Emit(decl_inst, MissingDefinitionInImpl);
+        }
+        break;
+      }
+      // TODO: handle `interface` as well, once we can test it without
+      // triggering https://github.com/carbon-language/carbon-lang/issues/4071
+      default: {
+        CARBON_CHECK(false);  // FIXME
+      }
```

Comment:
WDYT of being more explicit with this, such as:
```suggestion
      case CARBON_KIND(SemIR::InterfaceDecl interface_decl): {
        CARBON_FATAL() << "TODO: Add interfaces.";
      }
      default: {
        CARBON_FATAL() << "Unexpected inst in definitions_required: " << inst;
      }
```

Note that specific code would require storing the `inst` you have in the switch separately, we haven't made it available in CARBON_KIND_SWITCH (maybe should, just haven't).

Reply:
Good idea; done.

(And whoops, I should have caught that FIXME before sending)